julia/GradBench/src/GradBench.jl

```diff
@@ -2,48 +2,73 @@ module GradBench
 
 import JSON
 
-const DISPATCH_TABLE = Dict{String,Dict{String,Any}}()
+"""
+    Experiment
 
-function register!(mod, functions)
+Defines a measurment experiment.
+
+## Interface
+- `args = preprocess(::Experiment, message)`
+- `ret = (::Experiment)(args...)`
+- Optional: `postprocess(::Experiment, ret)`
+"""
+abstract type Experiment end
+
+function preprocess(::Experiment, message)
+    return message
+end
+
+function postprocess(::Experiment, ret)
+    return ret
+end
+
+const DISPATCH_TABLE = Dict{String,Dict{String,Experiment}}()
+
+function register!(mod, experiments)
     if haskey(DISPATCH_TABLE, mod)
         error("mod $mod is already registered")
     end
-    DISPATCH_TABLE[mod] = functions
+    DISPATCH_TABLE[mod] = experiments
     return
 end
 
 # Avoid measuring dispatch overhead
-function measure(func::F, arg) where {F}
+function measure(func::F, args...) where {F}
     start = time_ns()
-    ret = func(arg)
+    ret = func(args...)
     done = time_ns()
     return ret, done - start
 end
 
 function run(params)
     mod = DISPATCH_TABLE[params["module"]]
-    func = mod[params["function"]]
-    arg = params["input"]
+    experiment = mod[params["function"]]
     min_runs = get(params, "min_runs", 1)
     min_seconds = get(params, "min_seconds", 0)
     @assert min_runs > 0
 
-    # TODO: Prepare (parse JSON?)
-    # TODO: pre-allocate output?
     timings = Any[]
 
+    args, t = measure(preprocess, experiment, params["input"])
+    push!(timings, Dict("name" => "preprocess", "nanoseconds" => t))
```

Comment:
Judging by #480 this is called `"prepare"` for other tools?

Reply:
At least for the C++ tools, and there is no reason to diverge. In principle there is no limit to what fine-grained measurements a tool can provide (although the plots only care about `evaluate`).